cluster/raft_replication_apply_endpoints.go

```diff
@@ -14,15 +14,111 @@ package cluster
 import (
 	"context"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"strings"
 
 	"github.com/go-openapi/strfmt"
+	"github.com/google/uuid"
 	"github.com/weaviate/weaviate/cluster/proto/api"
 	"github.com/weaviate/weaviate/cluster/replication"
 	replicationTypes "github.com/weaviate/weaviate/cluster/replication/types"
 )
 
+func (s *Raft) ApplyReplicationScalePlan(ctx context.Context, scalePlan api.ReplicationScalePlan) (opsUUIDs []strfmt.UUID, err error) {
+	// while not strictly necessary, scaling while there are ongoing replications is disallowed
+	ops, err := s.GetReplicationDetailsByCollection(ctx, scalePlan.Collection)
+	if err != nil && !errors.Is(err, replicationTypes.ErrReplicationOperationNotFound) {
+		return nil, fmt.Errorf("get replication details for %q: %w", scalePlan.Collection, err)
+	}
+	for _, op := range ops {
+		if api.ShardReplicationState(op.Status.State) != api.CANCELLED && api.ShardReplicationState(op.Status.State) != api.READY {
+			return nil, fmt.Errorf("cannot scale while there are ongoing replications for collection %q", scalePlan.Collection)
+		}
+	}
+
+	// validate scale plan
+	for shardName, shardActions := range scalePlan.ShardReplicationScaleActions {
+		sourceNodeUsage := make(map[string]int)
+
+		for newNode, sourceNode := range shardActions.AddNodes {
+			if newNode == sourceNode {
+				return nil, fmt.Errorf("node %q in shard %q cannot be added as replica from itself", newNode, shardName)
+			}
+
+			if _, isBeingRemoved := shardActions.RemoveNodes[newNode]; isBeingRemoved {
+				return nil, fmt.Errorf("node %q in shard %q cannot be both removed and added", newNode, shardName)
+			}
+
+			if sourceNode != "" {
+				sourceNodeUsage[sourceNode]++
+				if sourceNodeUsage[sourceNode] > 1 {
+					if _, isBeingRemoved := shardActions.RemoveNodes[sourceNode]; isBeingRemoved {
+						return nil, fmt.Errorf("invalid scale plan: source node %q for shard %q is marked for removal and used multiple times as source for additions", sourceNode, shardName)
+					}
+				}
+			}
+		}
+	}
+
+	defer func() {
+		if err != nil {
+			for _, opsUUID := range opsUUIDs {
+				err := s.ForceDeleteReplicationByUuid(ctx, opsUUID)
```

Comment:
Why do we delete all the ops created so far instead of just the one that errored? Is the goal to do a best effort "all or nothing" scale in/out?

Reply:
If we don't delete the ops created previously, would the user be able to let the created ops finish, then recreate a new plan that would just finish whatever hasn't run yet